### PR TITLE
strip parentheses from the gpu product name

### LINKED ISSUE #514: cannot generate node label nvidia.com/gpu.product for GPU TITAN X
### Problem Brief

After deploy GFD and NFD in k8s cluster, I found that node with GPU TITAN X is missing label `nvidia.com/gpu.product: <gpu-product-name>` while node with GPU A100 contains label `nvidia.com/gpu.product: NVIDIA-A100-PCIE-40GB`.

### Environment

* NVIDIA GPU Driver: 525.125.06
* GFD version: v0.8.1
* NFD version: v0.10.1
* GPU Hareware: NVIDIA Corporation GP102 [TITAN X]

### Problem Details

GPU TITAN X's model name is `NVIDIA-TITAN-X-(Pascal)` 

GPU Feature Discovery generate labels as follow，`nvidia.com/gpu.product=NVIDIA-TITAN-X-(Pascal)` contains invalid character `(` & `)`: 

```bash
$ cat /etc/kubernetes/node-feature-discovery/features.d/gfd
nvidia.com/gpu.machine=HP-Z820-Workstation
nvidia.com/gpu.count=1
nvidia.com/gpu.memory=12288
nvidia.com/gpu.compute.major=6
nvidia.com/cuda.runtime.minor=0
nvidia.com/gpu.product=NVIDIA-TITAN-X-(Pascal)
nvidia.com/cuda.runtime.major=12
nvidia.com/mig.capable=false
nvidia.com/cuda.driver.minor=125
nvidia.com/cuda.driver.rev=06
nvidia.com/mig.strategy=single
nvidia.com/gpu.replicas=1
nvidia.com/gpu.compute.minor=1
nvidia.com/gpu.family=pascal
nvidia.com/gfd.timestamp=1692843500
nvidia.com/cuda.driver.major=525
```

Node Feature Discovery reject to create label `nvidia.com/gpu.product=NVIDIA-TITAN-X-(Pascal)` with following logs: 
```bash
$ k -n gpu-operator logs release-name-node-feature-discovery-worker-fjlcp 
I0813 01:26:04.719905       1 nfd-worker.go:472] starting feature discovery...
W0813 01:26:04.720702       1 nfd-worker.go:533] ignoring invalid feature value nvidia.com/gpu.product=NVIDIA-TITAN-X-(Pascal): [a valid label must be an empty string or consist of alphanumeric characters, '-', '_' or '.', and must start and end with an alphanumeric character (e.g. 'MyValue',  or 'my_value',  or '12345', regex used for validation is '(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])?')]
I0813 01:26:04.720752       1 nfd-worker.go:484] feature discovery completed
I0813 01:26:04.720767       1 nfd-worker.go:565] sending labeling request to nfd-master
```

### Conclusion

GFD should make sure generated labels are valid

### PR DESCRIPTION
**Pull request by tariq1890**
Fixes #514

### CODE REVIEW

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]
Could you add a small unit test for this addition please.

**Comment by tariq1890** [comment]
@elezar @ArangoGutierrez  Thank you for your review comments. I have addressed them. Requesting another round of review

File: internal/lm/resource_test.go [hunk 1]
```diff
@@ -216,6 +217,42 @@ func TestGPUResourceLabeler(t *testing.T) {
 
 }
 
+func TestGPUDeviceNameWithParentheses(t *testing.T) {
+	device := resource.DeviceMock{
```
Comment (ArangoGutierrez) [inline internal/lm/resource_test.go:221]: Since you also sanitize spaces `sanitised = strings.ReplaceAll(input, " ", "-")` let's add a test for that

**Review by ArangoGutierrez — CHANGES_REQUESTED** [review]

File: internal/lm/resource_test.go (hunk 1 above)
Comment (tariq1890) [reply internal/lm/resource_test.go:221]: Sure

Comment (tariq1890) [reply internal/lm/resource_test.go:221]: The reason I didn't add it was because it was already covered in previous test cases

**Review by ArangoGutierrez — APPROVED** [review]

File: internal/lm/resource.go [hunk 2]
```diff
@@ -187,7 +187,11 @@ func (rl resourceLabeler) getProductName(parts ...string) string {
 	var strippedParts []string
 	for _, p := range parts {
 		if p != "" {
-			strippedParts = append(strippedParts, strings.ReplaceAll(p, " ", "-"))
+			var strippedPart string
+			strippedPart = strings.ReplaceAll(p, " ", "-")
+			strippedPart = strings.ReplaceAll(strippedPart, "(", "")
+			strippedPart = strings.ReplaceAll(strippedPart, ")", "")
+			strippedParts = append(strippedParts, strippedPart)
```
[the remark this answers is not in the document]
Comment (tariq1890) [reply internal/lm/resource.go]: This is done

File: internal/lm/resource_test.go (hunk 1 above)
Comment (elezar) [reply internal/lm/resource_test.go:221]: We can add:
```
func TestResourceLabeler(t *testing.T) {
	testCases := []struct {
		productParts        []string
		expectedProductName string
	}{
		{
			productParts:        []string{"some", "part"},
			expectedProductName: "some-part",
		},
		{
			productParts:        []string{"some", "", "more other", "parts"},
			expectedProductName: "some-more-other-parts",
		},
	}

	for i, tc := range testCases {
		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
			rl := resourceLabeler{}

			require.EqualValues(t, tc.expectedProductName, rl.getProductName(tc.productParts...))
		})
	}
}
```
to test other combinations and the interactions with the sharing config.

**Review by elezar — COMMENTED** [review]
See below on testing.

File: internal/lm/resource_test.go (hunk 1 above)
Comment (tariq1890) [reply internal/lm/resource_test.go:221]: I can we can also do a `TrimWhiteSpaces` to further sanitise the string

**Comment by tariq1890** [comment]
Thanks @elezar ! I've updated the logic and it passes all the test cases you've provided in the review comments

File: internal/lm/resource_test.go [hunk 3]
```diff
@@ -216,6 +216,36 @@ func TestGPUResourceLabeler(t *testing.T) {
...
+			expected: "somethingelse",
+		},
+		{
+			input:    "some ( thing )else",
+			expected: "some-thing-else",
+		},
+		{
+			input:    "NVIDIA-TITAN-X-(Pascal)",
+			expected: "NVIDIA-TITAN-X-Pascal",
+		},
```
Comment (elezar) [inline internal/lm/resource_test.go:239]: Not a blocker, but adding the following explicitly may be good.
```suggestion
		},
		{
				input: " input  with multiple   spaces   ",
				expected: "input-with-multiple-spaces",
		},
```

File: internal/lm/resource.go [hunk 4]
```diff
@@ -302,3 +303,15 @@ func getArchFamily(computeMajor, computeMinor int) string {
...
+
+func sanitise(input string) string {
+	var sanitised string
+	sanitised = strings.ReplaceAll(input, "(", "")
+	sanitised = strings.ReplaceAll(sanitised, ")", "")
+
+	// remove redundant blank spaces
+	sanitised = strings.Join(strings.Fields(strings.TrimSpace(sanitised)), " ")
+
+	sanitised = strings.ReplaceAll(sanitised, " ", "-")
```
Comment (elezar) [inline internal/lm/resource.go:315]: Looking at `strings.Fields` https://pkg.go.dev/strings#Fields I think that the `strings.TrimSpace` is redundant. We could simplify the following to:
```suggestion
	sanitised = strings.Join(strings.Fields(sanitised), "-")
```

**Review by elezar — CHANGES_REQUESTED** [review]
Thanks @tariq1890.

I have a suggestion for a minor optimisation. With that in we should be good!

File: internal/lm/resource.go (hunk 4 above)
Comment (tariq1890) [reply internal/lm/resource.go]: Done, thanks

File: internal/lm/resource_test.go (hunk 3 above)
Comment (tariq1890) [reply internal/lm/resource_test.go:239]: Done

**Review by elezar — APPROVED** [review]

**Pull request merged**